### ISSUE TITLE
Problem with boringSSL when making an http2 server

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#24208]
Making a http2 server using a dartsdk with boringssl support fails (whereas using an older sdk with NSS succeeds) to establish a connection with a chrome/firefox client.

The error messages are:
```
139993520541664:error:100b60c1:SSL routines:ssl3_get_client_hello:NO_SHARED_CIPHER:third_party/boringssl/src/ssl/s3_srvr.c:1085:

Unexpected error: HandshakeException: Handshake error in server (OS Error: Unknown error from BoringSSL library, errno = 0)
Unexpected error - stack: #0      _SecureFilterImpl.handshake (dart:io-patch/secure_socket_patch.dart:94)
#1      _RawSecureSocket._secureHandshake (dart:io/secure_socket.dart:778)
#2      _RawSecureSocket._tryFilter.<anonymous closure> (dart:io/secure_socket.dart:884)
#3      _rootRunUnary (dart:async/zone.dart:910)
#4      _CustomZone.runUnary (dart:async/zone.dart:810)
#5      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:502)
#6      _Future._propagateToListeners (dart:async/future_impl.dart:585)
#7      _Future._completeWithValue (dart:async/future_impl.dart:376)
#8      _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:430)
#9      _rootRun (dart:async/zone.dart:903)
#10     _CustomZone.run (dart:async/zone.dart:802)
#11     _CustomZone.runGuarded (dart:async/zone.dart:708)
#12     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:733)
#13     _microtaskLoop (dart:async/schedule_microtask.dart:43)
#14     _microtaskLoopEntry (dart:async/schedule_microtask.dart:52)
#15     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#16     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)

139993520541664:error:100b60c1:SSL routines:ssl3_get_client_hello:NO_SHARED_CIPHER:third_party/boringssl/src/ssl/s3_srvr.c:1085:

Unexpected error: HandshakeException: Handshake error in server (OS Error: Unknown error from BoringSSL library, errno = 0)
Unexpected error - stack: #0      _SecureFilterImpl.handshake (dart:io-patch/secure_socket_patch.dart:94)
#1      _RawSecureSocket._secureHandshake (dart:io/secure_socket.dart:778)
#2      _RawSecureSocket._tryFilter.<anonymous closure> (dart:io/secure_socket.dart:884)
#3      _rootRunUnary (dart:async/zone.dart:910)
#4      _CustomZone.runUnary (dart:async/zone.dart:810)
#5      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:502)
#6      _Future._propagateToListeners (dart:async/future_impl.dart:585)
#7      _Future._completeWithValue (dart:async/future_impl.dart:376)
#8      _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:430)
#9      _rootRun (dart:async/zone.dart:903)
#10     _CustomZone.run (dart:async/zone.dart:802)
#11     _CustomZone.runGuarded (dart:async/zone.dart:708)
#12     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:733)
#13     _microtaskLoop (dart:async/schedule_microtask.dart:43)
#14     _microtaskLoopEntry (dart:async/schedule_microtask.dart:52)
#15     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#16     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)

139993520541664:error:100dd0a3:SSL routines:ssl_bytes_to_cipher_list:INAPPROPRIATE_FALLBACK:third_party/boringssl/src/ssl/ssl_lib.c:1

Unexpected error: HandshakeException: Handshake error in server (OS Error: Unknown error from BoringSSL library, errno = 0)
Unexpected error - stack: #0      _SecureFilterImpl.handshake (dart:io-patch/secure_socket_patch.dart:94)
#1      _RawSecureSocket._secureHandshake (dart:io/secure_socket.dart:778)
#2      _RawSecureSocket._tryFilter.<anonymous closure> (dart:io/secure_socket.dart:884)
#3      _rootRunUnary (dart:async/zone.dart:910)
#4      _CustomZone.runUnary (dart:async/zone.dart:810)
#5      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:502)
#6      _Future._propagateToListeners (dart:async/future_impl.dart:585)
#7      _Future._completeWithValue (dart:async/future_impl.dart:376)
#8      _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:430)
#9      _rootRun (dart:async/zone.dart:903)
#10     _CustomZone.run (dart:async/zone.dart:802)
#11     _CustomZone.runGuarded (dart:async/zone.dart:708)
#12     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:733)
#13     _microtaskLoop (dart:async/schedule_microtask.dart:43)
#14     _microtaskLoopEntry (dart:async/schedule_microtask.dart:52)
#15     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#16     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)
```

**Comment by whesse** [comment]
After the fix in 891c36a8e03399dcb4ef53e20b2fa0ef5260b51a , an example HTTP server will connect to Chrome and Firefox, and if the certificate is added temporarily to the trusted roots (DO NOT LEAVE IT THERE - security fail), a web page can be fetched.  The errors you see disappear.  If ALPN protocols are added, then if 'http/1.1' is first, it succeeds, and if 'h2' is first, it fails (as expected).   So you should have a good start to making it work now.  Here is the sample server, that works with Chrome and FF:

String localFile(path) => Platform.script.resolve(path).toFilePath();
SecurityContext serverContext = new SecurityContext()
  ..useCertificateChain(localFile('certificates/server_chain.pem'))
  ..usePrivateKey(localFile('certificates/server_key.pem'),
                  password: 'dartdart')
  ..setAlpnProtocols(['http/1.1', 'foo'], true)
  ..setAlpnProtocols(['h2', 'http/1.1'], true)  // Does not work.
    ;

  void test() {
    HttpServer.bindSecure(HOST,
                          0,
                          serverContext,
                          backlog: 5).then((server) {
      print("Https server on port ${server.port}");
      server.listen((HttpRequest request) {
            print("request received");
            request.response.write("hello world");
            request.response.close();
      });
    });
  }

void main() {
  InternetAddress.lookup("localhost").then((hosts) {
    HOST = hosts.first;
    test();
  });
}

**Comment by mkustermann** [comment]
Thanks Bill. I can confirm that the exceptions are gone now.
